### PR TITLE
CM-145 How to Use Local functions in C#

### PR DESCRIPTION
**Pull request by shayar**

### CODE REVIEW

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿namespace LocalFunctionInCSharp
+{
+    public class Product
+    {
+        public string Name { get; set; }
+        public int StockQuantity { get; set; }
+        public int SelectedQuantity { get; set; }
+
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs:8]: Extra newline here

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+﻿namespace LocalFunctionInCSharp
+{
+    public class ShoppingCart
+    {
+        public List<Product> Products { get; set; }
+        public bool Checkout()
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs:6]: Need a blank line here

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs (hunk 1 above)
Comment (shayar) [reply csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs]: Removed it.

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs (hunk 2 above)
Comment (shayar) [reply csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs:7]: Added

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿namespace LocalFunctionInCSharp
+{
+    public class Product
+    {
+        public string Name { get; set; }
+        public int StockQuantity { get; set; }
+        public int SelectedQuantity { get; set; }
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs:7]: All of these properties could be made init only, it is also a good candidate to be a record based on its usage.

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+﻿using LocalFunctionInCSharp;
+
+var cart = new ShoppingCart
+{
+    Products = new List<Product>
+            {
+                new Product { Name = "Product A", StockQuantity = 10, SelectedQuantity = 2 },
+                new Product { Name = "Product B", StockQuantity = 11, SelectedQuantity = 5 }
+            }
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs:9]: Use target typed new

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+if (cart.Checkout())
+{
+    Console.WriteLine("Checkout successful.");
+}
+else
+{
+    Console.WriteLine("Checkout failed due to validation errors.");
+}
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs:19]: We could rewrite this with the ternary operator:
`Console.WriteLine(cart.Checkout() ? "Checkout successful." : "Checkout failed due to validation errors.");`

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+List<int> numbers = new List<int> { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+List<int> evenNumbersUsingLocalFunction = FilterEvenNumbersUsingLocalFunction(numbers);
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs:22]: Use var (there are several instances of this in the code)

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+Console.WriteLine(string.Join(", ", evenNumbersUsingLocalFunction));
+
+static List<int> FilterEvenNumbersUsingLocalFunction(List<int> numbers)
+{
+    return numbers.Where(IsEven).ToList();
+
+    bool IsEven(int num)
+    {
+        return num % 2 == 0;
+    }
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs:32]: Use expression body syntax:
`bool IsEven(int num) => num % 2 == 0;`

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+
+List<int> evenNumbersUsingLambdaFunction = FilterEvenNumbersUsingLambda(numbers);
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Program.cs:36]: Use var

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
+﻿namespace LocalFunctionInCSharp
+{
+    public class ShoppingCart
+    {
+        public List<Product> Products { get; set; }
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs:5]: Can be init only and should be marked required

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿namespace LocalFunctionInCSharp
+{
+    public class Product
+    {
+        public string Name { get; set; }
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/Product.cs:5]: Should be marked required

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+            bool IsQuantityValid(Product product, int quantity)
+            {
+                return quantity > 0 && quantity <= product.StockQuantity;
+            }
+
+            void LogValidationFailure(string message)
+            {
+                Console.WriteLine($"Validation Error:- {message}");
+            }
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs:37]: Local functions should come after the return value

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
+using LocalFunctionInCSharp;
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs:1]: Move to global using

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        public void GivenLessSelectedQuantityThanStockQuantity_ThenSuccessfulCheckout()
+        {
+            var cart = new ShoppingCart
+            {
+                Products = new List<Product>
+            {
+                new Product { Name = "Product A", StockQuantity = 10, SelectedQuantity = 2 },
+                new Product { Name = "Product B", StockQuantity = 5, SelectedQuantity = 3 }
+            }
+            };
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs:17]: Formatting here is off, also we can use target typed new

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+            bool result = cart.Checkout();
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs:19]: Use var

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+        public void GivenMoreSelectedQuantityThanStockQuantity_ThenProductOutOfStockFails()
+        {
+            var cart = new ShoppingCart
+            {
+                Products = new List<Product>
+            {
+                new Product { Name = "Product A", StockQuantity = 0, SelectedQuantity = 2 },
+                new Product { Name = "Product B", StockQuantity = 5, SelectedQuantity = 3 }
+            }
+            };
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs:34]: Spacing/indentation is off also we can use target typed new

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+            bool result = cart.Checkout();
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs:36]: Use var

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+            var cart = new ShoppingCart
+            {
+                Products = new List<Product>
+            {
+                new Product { Name = "Product A", StockQuantity = 10, SelectedQuantity = 20 },
+                new Product { Name = "Product B", StockQuantity = 5, SelectedQuantity = -1 }
+            }
+            };
+
+            bool result = cart.Checkout();
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs:53]: See earlier comments

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+            Assert.False(result);
+        }
+
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharpTests/ShoppingCartTests.cs:57]: Extra blank line

File: csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+            return true;
+        }
+    }
+}
+
+
```
Comment (jshergal) [inline csharp-basic-topics/LocalFunctionInCSharp/LocalFunctionInCSharp/ShoppingCart.cs:44]: Extra blank lines

**Review by jshergal — COMMENTED** [review]
It would also reduce nesting if we switched to file scoped namespaces, which is the default setting in VS for new .NET 7.0 projects.

**Pull request closed without merging**